### ISSUE TITLE
Ranking Games and new Lobby window

### ISSUE THREAD

**Issue by xan2622** [OpenRA/OpenRA#2815]
Hello.

Multiplayer games could become even more interesting if a Ranking system was implemented.

Concept:
- Players who want their game to be part of the ranking would just have to enable a checkbox during the game creation
- Winners of ranking games would be "recorded" and results could be sent to a database (on the website)
- The best players would see their ranking rise and their names listed in a table, on the website
- The ranking could last one month but there could be statistics for 3 months, 6 months, a year...
- Over a season, each player would be able to know his personal statistics such as: 
  number of games won, number of games lost, number of games played, pourcentage of 
  games won, the fastest game won..

To support this new Ranking system, the lobby would have to be changed a bit.
It should feature options to filter games, find players easily, ...

Here is a mockup:
![607690newlobby1](https://f.cloud.github.com/assets/1732773/311051/1d192310-972f-11e2-9388-5d7ccb2dd111.jpg)
- The lobby could be upgraded to allow chatting with all connected players (in a global chatroom)
- There would also be a drop-down menu to filter games by OpenRA version, by mod, by type (standard games, private games (invite-only), ranking game)
  Mockup: 
  ![630164newlobby2](https://f.cloud.github.com/assets/1732773/311052/25239c0c-972f-11e2-9aea-7a5d0d38c22e.jpg)
- The Players list (on the left) could also be sorted
  Mockup:
  ![787501newlobby3](https://f.cloud.github.com/assets/1732773/311053/2cb31600-972f-11e2-983d-5d3a917c6f6a.jpg)
- A new button could be added to randomly join any open game

**Comment by jwbuurlage** [comment]
@clemby Are you still working on this? I would like to take this on.

**Comment by x-a-n-a-x** [comment]
@jwbuurlage : how do you plan to implement this?

With an ingame ladder? 
With a third party ladder like http://binarybeast.com/api/info ?

**Comment by jwbuurlage** [comment]
@x-a-n-a-x My initial thought was to have a statistics server, and let game servers (optionally) register finished games there (ranked or unranked alike) using some basic RESTful API. This could be a implemented as a server option that defaults to false. This "statistics server" would allow for multiple output formats for use by an ingame ladder and external websites.  This is not unlike how the master server is set up.

For example, a simple API call could be used to obtain current (paged) rankings, for easy integration. I suppose the statistics gathered for a game should be as extensive as possible for future use, but the initial version would only be used for an ELO ladder of some kind. Later the same information could be used for player profiles etc.

If no one is working on this currently, I can draft a more detailed plan.

**Comment by GeertJohan** [comment]
I'm also working on something like this. I have plans for a web-based multiplayer platform for openRA. The first version will be very simple, just have a server list and a general chat. But in the end I would like to have an account system, ranked matches, matchmaking, etc.

**Comment by jwbuurlage** [comment]
@GeertJohan Maybe it would be a good idea to collaborate on this with everyone interested, it seems like a lot of people have tried and gave up. Why would we make a web-based platform? An in-game general lobby and a ranked game list would suffice, and we would not be segregating the player-base. Maybe we can split this into a number of different tasks:

- [ ] Authentication system [#2265,  #4410]
- [ ] General chat (could simply connect to #openra)
- [ ] Statistics server
- [ ] Ranked matches / Ladder
- [ ] Automatic matchmaking

**Comment by clemby** [comment]
@jwbuurlage still working on it, but don't let that stop you if you have better ideas.

I've been working towards adding a multiplayer ladder to the [tournament site](http://compo.openra.net), extending its API and adding appropriate views etc. There is still a lot left though: I haven't implemented the ranking logic or made any changes to the game itself yet (lobby and ranking checkbox as described above); I'm also undecided on how best to authenticate players. Help (anywhere but especially in these areas) is welcome too.  
Some notes:
- An Elo system may be less appropriate for a ladder based on multiplayer games than other ranking systems, such as [Glicko-2](http://www.glicko.net/glicko.html); alternatively it might be fun to create our own.
- If unranked games can be recorded, I think that should be the player's decision; I wouldn't want to discourage more relaxed games.

**Comment by Greg-Boggs** [comment]
Elo and Glicko-2 are both well suited for online games. The problems with both are similar. Namely, both formulas require 100% random matching between players. If players can pick their matches, it's trivial to inflate rankings in both systems. 

On the Wesnoth Ladder, we made more than a few tweaks to improve this. First, new players give out half points. This discourages spam accounts and targeting of newbies for easy points. Second, we've adjusted the point system to allow players to climb over time regardless of their skill. The reason for that is players don't like realizing their rank doesn't increase. It's just not as fun as making small incremental improvements with effort.

On the topic of match recording and integration, the Wesnoth Ladder works entirely with no integration. All matches are player reported and uploaded replays are done manually by the players. It works wonderfully and required significantly less programming to do.

**Comment by jwbuurlage** [comment]
@Greg-Boggs Very interesting, especially your last paragraph. Can any replay be submitted? Do both parties have to submit the replay? I can see the following problems with this approach, curious how you guys resolved this:
- What if a casual game is played, and one party decides to upload the recording solely on the basis that he has won and wants to improve his ranking? Or more generally, how do you handle ranked versus unranked games?
- What stops a player from not reporting a game if he loses? Or is one-party consent enough?
- Wouldn't it be discouraging for olders players to play new players if they only receive half points, or would they only lose half points as well? In other games I have seen ranking systems where experienced players playing new player stood to gain very little points when they won, but received massive penalties when they would lose. 
- Do you guys have any random matchmaking in place? 
- About your second tweak, can you give some more details about your implementation of this?

**Comment by Greg-Boggs** [comment]
You should sign up for an account and look at the report page: http://wesnoth.gamingladder.info/report.php

Results and replays are submitted by the match winner. Only mutually agreed ranked games are reported. It's possible for people to cheat. But, actual cheating is very, very low and the stats of reversed games are published on the home page. It does take 1 person willing to be the moderator to settle these infrequent incidents. 

The goal of newbie protection was to discourage top players from picking on newbies. The number of games to remove protection are very few. A couple newbies playing together for 2-3 hours could get out of protection.

Wesnoth does not support randomized matches. We requested the feature, but the core team doesn't want to see competitor play built directly into Wesnoth. 

I believe the place to look for implementation details are in the constants file of the project: http://sourceforge.net/projects/gamingladder/files/ 

Essentially, we tweaked the number of points gained so that people gain slightly more than they lose. One commonly requested feature is a second ladder where the ladder resets on a periodic basis while still maintaining the long term standings.

**Comment by MicroJOo** [comment]
@jwbuurlage , @Greg-Boggs  : you guys may find this here interesting :  #3859

@ScottNZ made some approach here (with some outstanding results) , but at the end it got removed ... see here :  #6786 ... as far as i remember , plan was to build our ranking system on top of that implemented irc chat

**Comment by GeertJohan** [comment]
@jwbuurlage and I have teamed up and we are actively working on this.

**Comment by x-a-n-a-x** [comment]
http://www.sleipnirstuff.com/forum/viewtopic.php?f=82&t=17810&start=0&postdays=0&postorder=asc&highlight=

**Comment by pchote** [comment]
The MP lobby has been overhauled and now looks much like these mockups.  Rankings first requires authentication, and so this is now largely a duplicate of #2265 / #9990.

**Comment by vizcay** [comment]
Hi! what's the current status of this!?

I was thinking to colaborate, being a full stack developer I can develop the API & simple web interface (twitter bootstrap) for a ranking system like ELO (I also play chess), with profiles, very fast. I've read about implementing this with IRC and I think it will lack flexibility in the future.

But I'm not a C# developer so I wont be able to submit pull request for the main game, only develop the server side.

Stack: Ruby On Rails, PostgresSQL, Nginx, Linux.

**Comment by pchote** [comment]
This is still blocked on having a player authentication system that works with the game.  IRC was previously discussed as an option for that, but I personally think a much better approach would be to integrate it with our planned new forum (as a phpBB3 modification).

There's really no point in implementing rankings until we can authenticate that players are who they claim they are.  It is far too prone to abuse otherwise.

**Comment by vizcay** [comment]
Hi @pchote ,
authentication is central to the ranking system of course.. but, server side speaking, is not difficult at all, and I don't believe client side with C# to be too troublesome..

Basically,
1. Player browses to ranking site, lets say http://league.openra.net and creates an account.
2. It's inserts it's login name and password into OpenRA.
3. When the player opens OpenRA multiplayer lobby it's gets authenticated by connecting to API endpoint, of course we never send plain passwords over the wire.. a simple scheme like HTTP Digest Authentication can be used.
4. API endpoint responds with a token with a limited lifetime (or else the authentication process has to be done every time the client comunicates with the API).
5. When a game starts both clients contact the API informing that it will be a ranked game.
6. When a game finishs, the client contact the API informing the result.

Of course there a few things to consider like what will happen if both clients inform different results (cheating?), etc. But for starting is not difficult at all, I bet I can have a prototype working in like a week.

**Comment by pchote** [comment]
If you wanted to put some time in to build a rough backend prototype that could be used as a starting point for the client side work, then that would be great.  Once the client and server sides are working together we could decide on how/if we want to proceed further.

The standing plan for handling ranked games and avoiding cheating is to have a set of whitelisted ranked servers, and then the servers report the game results to the backend.  If players want to play unranked games then they can play on unranked servers.  We would then have API methods for clients to verify that a ranked server is what it claims to be, and for the server to verify that clients are who they claim to be.

**Comment by vizcay** [comment]
@pchote:
I understand your concerns about the whitlisting of servers.. but this is not a profesional league (aka starcraft in south corea) or something that needs ultra security, like homebanking. And yet it will add a lot to the comunity in terms of new players and traffic. Also whitelisted servers can add a bottleneck if trafic starts to grow.

What about this scheme:
1- When a ranked game starts both clients comunicate separately to the server via the API (this is not repudiable if implemented well).
2- When a ranked game ends both clients comunicate the result. If both comunicate the same, all good. If there is a discrepancy, well there will be the need of manual adjustment and banning of the ofending player.
3- Disconnects losts after X time or disconnect cancels game policy.

The worst thing I can think of is the game server disconnecting the client on purpose to claim a disconnect...

But,
1- it's very easy to detect a cheat (manually, checking the replay), and an ip ban can also be applied.
2- it's very difficult to abuse the system in the long run and on which incentives? (not money to win).
3- not everyone is tech-savvy enought to hijack even simple schemes like this.

By the way, are u currently using XML or JSON as the payload of the HTTP API ?

**Comment by pchote** [comment]
The server browser and lobby windows have changed a lot since this was created, and we have other issues covering the specifics of authentication and ranking.  Closing.
